### PR TITLE
 Support CNI 0.3.x in openshift-sdn

### PR DESCRIPTION
**Pull request by luksa**
This PR makes openshift-sdn support CNI version 0.3.0 and 0.3.1, which allows openshift-sdn to be used with additional CNI plugins (plugin chaining). 

Fixes https://github.com/openshift/origin/issues/21891

### CODE REVIEW

**Comment by luksa** [comment]
@soltysh @knobunc Can someone review this? The Istio team needs this.

**Comment by mabdelgaid** [comment]
Can anyone take care of this one?? we have a customer looking for this option for security reasons

**Comment by dcbw** [comment]
@luksa I'd rather see the plugin just create a 0.3.1 result from the start with current.Result{}. No point in creating the 0.2.0 result just to convert it to 0.3.1.

**Comment by dcbw** [comment]
@luksa You can find some examples of the current.Result construction here:

https://github.com/containernetworking/plugins/blob/master/plugins/main/macvlan/macvlan.go#L174

pkg/network/node/pod.go needs an update in ipamAdd() to call current.NewResultFromResult() on the result that ipam.ExecPluginWithResult() returns. That will eventually get back to openshift-sdn_linux.go in doCNIServerAdd() at which point you should do the same thing to create a current.Result() from whatever comes back from doCNI(), instead of the types020.NewResult() call it currently does.

We shouldn't really need convertTo03x() at all.

Does all that make sense?

**Comment by luksa** [comment]
@dcbw Yup, made perfect sense. I had seen that code before, I just wasn't sure if I should be modifying the server part or not.

I've now changed this per your instructions, but I also had to change the CNI version used in the IPAM plugin call, because using `current.NewResultFromResult` also invokes the broken convert function (which would again require me to include the fixed `convertTo03x()` function). Without this final change, it didn't work (because the gateway was set on the routes instead of being nil). 

Anyway, by invoking the IPAM plugin with version 0.3.1 and directly reading its 0.3.1 result, there's no need for any conversion any more, and everything appears to work properly.

File: cmd/sdn-cni-plugin/openshift-sdn_linux.go [hunk 1]
```diff
@@ -94,25 +93,33 @@ func (p *cniPlugin) doCNI(url string, req *cniserver.CNIRequest) ([]byte, error)
...
+	return result.(*current.Result), nil
 }
 
 func (p *cniPlugin) testCmdAdd(args *skel.CmdArgs) (types.Result, error) {
-	return p.doCNIServerAdd(newCNIRequest(args), "dummy0")
+	result, err := p.doCNIServerAdd(newCNIRequest(args), "dummy0")
+	if err != nil {
+		return nil, err
+	}
+	convertedResult, err := convertToRequestedVersion(args.StdinData, result)
```
Comment (dcbw) [inline cmd/sdn-cni-plugin/openshift-sdn_linux.go]: Just do `return convertToRequestedVersion(args.StdinData, result)` here; no need for the other 4 lines if we're not doing any additional error processing or touching convertedResult.

File: cmd/sdn-cni-plugin/openshift-sdn_linux.go [hunk 2]
```diff
@@ -234,7 +236,30 @@ func (p *cniPlugin) CmdAdd(args *skel.CmdArgs) error {
 		return err
 	}
 
-	return result.Print()
+	result.Interfaces = nil
+	result.IPs[0].Interface = nil
+	result.IPs[0].Gateway = defaultGW
```
Comment (dcbw) [inline cmd/sdn-cni-plugin/openshift-sdn_linux.go]: I wouldn't do this; just leave the Interfaces and IPs array alone and correct for a 0.3.x Result type. If the runtime requested a different version (eg 0.1 or 0.2) then convertToRequestedVersion() will handle that downconvert.

Comment (luksa) [reply cmd/sdn-cni-plugin/openshift-sdn_linux.go]: Done.

File: cmd/sdn-cni-plugin/openshift-sdn_linux.go (hunk 1 above)
Comment (luksa) [reply cmd/sdn-cni-plugin/openshift-sdn_linux.go]: Fixed. Thanks.

**Comment by luksa** [comment]
@dcbw re-review plz? :)

**Comment by mangelajo** [comment]
:+1 Ok, I have tested this side by side with another cni chained and it works!

**Pull request merged**